### PR TITLE
Wip rebuild monstore

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: http://tracker.ceph.com/issues/17179

### CODE REVIEW

File: src/tools/CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +33,7 @@ install(PROGRAMS
 
 add_executable(ceph-objectstore-tool
   ceph_objectstore_tool.cc
+  rebuild_mondb.cc
```
Comment (athanatos) [inline src/tools/CMakeLists.txt]: Probably also needs to go in automake?

**Comment by athanatos** [comment]
@gregsfortytwo Can you take a look?

File: src/tools/CMakeLists.txt (hunk 1 above)
Comment (tchaikov) [inline src/tools/CMakeLists.txt]: will do. thought we were ditching off automake =)

File: doc/rados/troubleshooting/troubleshooting-mon.rst [hunk 2]
```diff
@@ -383,6 +383,80 @@ example::
...
+---------------------------------
+
+If there is any survivers, we can always `replace`_ the corrupted one with a
+new one. And after booting up, the new joiner will sync up with a healthy
+peer, and once it is fully sync'ed, it will be able to serve the clients.
+
+Recovery using OSDs
+-------------------
+
+But what if all monitors fails at the same time? Since users are encouraged to
```
Comment (tchaikov) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: s/fails/fail/

File: src/tools/ceph_monstore_tool.cc [hunk 3]
```diff
@@ -509,6 +514,148 @@ int inflate_pgmap(MonitorDBStore& st, unsigned n, bool can_be_trimmed) {
   return 0;
 }
 
+static int update_auth(MonitorDBStore& st, const string& keyring_path)
+{
+  // add a almighty ceph.client.admin.keyring
```
Comment (tchaikov) [inline src/tools/ceph_monstore_tool.cc]: s/a/an/

**Comment by gregsfortytwo** [comment]
Paging @jecluis as well. ;)

File: doc/rados/troubleshooting/troubleshooting-mon.rst [hunk 4]
```diff
@@ -383,6 +383,80 @@ example::
 
 	iptables -A INPUT -m multiport -p tcp -s {ip-address}/{netmask} --dports 6789,6800:7300 -j ACCEPT
 
+Monitor Store Failures
+======================
+
+Symptoms of store corruption
+----------------------------
+
+Ceph monitor stores the `cluster map`_ in a key/value store. If a monitor in a
```
Comment (gregsfortytwo) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: We might want to say something like "a key/value store such as LevelDB".

I realize it's configurable and we might switch to RocksDB, but I think in the medium term anybody looking at this is going to know that LevelDB got messed up and not seeing that this is a repair tool for that case will confuse them. (Avoiding the name is especially weird since we are including LevelDB error messages below.)

File: doc/rados/troubleshooting/troubleshooting-mon.rst [hunk 5]
```diff
@@ -383,6 +383,80 @@ example::
...
+If there is any survivers, we can always `replace`_ the corrupted one with a
+new one. And after booting up, the new joiner will sync up with a healthy
+peer, and once it is fully sync'ed, it will be able to serve the clients.
+
+Recovery using OSDs
+-------------------
+
+But what if all monitors fails at the same time? Since users are encouraged to
+deploy at least three monitors in a Ceph cluster, the chance of simultaneous
+failure is rare. But unplanned power-downs in a data centre could fail the
```
Comment (gregsfortytwo) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: How about "in a data center with improperly configured disk/fs settings"; we don't want to make it sound random.

File: doc/rados/troubleshooting/troubleshooting-mon.rst [hunk 6]
```diff
@@ -383,6 +383,80 @@ example::
...
+underlying filesystem, and hence kill all the monitors. In this case, we can
+recover the monitor store with the information stored in OSDs.::
+
+  ms=/tmp/mon-store
+  mkdir $ms
+  # collect the cluster map from OSDs
+  for host in $hosts; do
+    rsync -avz /tmp/mon-store user@host:$ms
```
Comment (gregsfortytwo) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: Why do we want to send over our existing maps to the remote side first?

Oh, are we sending all ours, then grabbing theirs, then overwriting everything we have locally? Why not just use rsync to merge them?

Comment (athanatos) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: Is that even possible?  Even if it unnecessarily moves stuff around, I think this is a pretty clear explanation of what needs to be where and how the tools are meant to be used, which is what's really important.

Comment (athanatos) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: @tchaikov Oh, what happens if one of the OSDs happens to be really old and there is a hole between its newest map and the oldest map on the others?  I think the correct answer is for the mon to have the newest contiguous chunk of OSDMaps, but I'm not sure if you handled that.

**Comment by gregsfortytwo** [comment]
Still reading code, but I don't see any tests for this. I just *know* it's going to break every time we change storage stuff (like with ceph-mgr or something, for instance) and we aren't going to notice if the nightlies don't tell us.

So a basic scenario where we append a "rebuild" task at the end of one (or several) normal RADOS tests to start with. At least one where we have an OSD we deliberately keep out-of-date like Sam mentioned above. One we'll write when the FS repair stuff is done. Perhaps others but that's all I've got off the top of my head.

File: doc/rados/troubleshooting/troubleshooting-mon.rst (hunk 6 above)
Comment (gregsfortytwo) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: Oh, this is actually creating a MonDBStore in place; I'd assumed it would be extracting to some intermediate format and then turning them into the mon format at the end. So yes, we definitely need to copy it back and forth to do that.

File: src/tools/rebuild_mondb.cc [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+// rebuild
+//  - auth/${epoch}
+//  - auth/first_committed
+//  - auth/last_committed
+static int update_auth(const string& keyring_path,
+                       const OSDSuperblock& sb,
+                       MonitorDBStore& ms)
+{
+  // stolen from AuthMonitor::prepare_command(), where prefix is "auth add"
```
Comment (gregsfortytwo) [inline src/tools/rebuild_mondb.cc]: It sounds like you actually copied the functions? If so we should definitely pull them out into a shared cc file or something, rather than having two copies that can diverge with different behavior or bugs!

File: src/tools/rebuild_mondb.cc [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+    ntrimmed++;
+  }
+  if (sb.oldest_map > first_committed) {
+    t->put(prefix, first_committed_name, sb.oldest_map);
+  }
+
+  // add inc maps
+  unsigned nadded = 0;
+  const auto last = max(last_committed+1, sb.oldest_map);
+  for (auto e = last; e <= sb.newest_map; e++) {
```
Comment (gregsfortytwo) [inline src/tools/rebuild_mondb.cc]: So here, if the OSD has newer maps but an out-of-date newest_map, we're leaving them behind. Can we do a listing of the prefix range and pull them all out instead?

File: src/tools/rebuild_mondb.cc [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+    int nread = fs.read(coll_t::meta(), oid, 0, 0, bl);
+    if (nread <= 0) {
+      cerr << "missing " << oid << std::endl;
+      return EINVAL;
+    }
+    t->put(prefix, e, bl);
+    nadded++;
+  }
+  // add full maps
+  for (auto e = last; e <= sb.newest_map; e++) {
```
Comment (gregsfortytwo) [inline src/tools/rebuild_mondb.cc]: Same here.

File: src/tools/rebuild_mondb.cc [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+  // trim stale maps
+  unsigned ntrimmed = 0;
+  for (auto e = first_committed; e && e < sb.oldest_map; e++) {
+    t->erase(prefix, e);
+    t->erase(prefix, ms.combine_strings("full", e));
+    ntrimmed++;
+  }
+  if (sb.oldest_map > first_committed) {
+    t->put(prefix, first_committed_name, sb.oldest_map);
+  }
```
Comment (gregsfortytwo) [inline src/tools/rebuild_mondb.cc]: Oh wait, if we do have a discontiguous OSD that we read first, and then later we chop off its maps, I think we leak the old ones in our store? Not a huge deal, but we should set up a trim if we run into this case.

**Comment by gregsfortytwo** [comment]
Otherwise looks fine to me (although I only skimmed the pgmap updates).

File: src/tools/rebuild_mondb.cc (hunk 7 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: not really, i adapted them.

File: src/tools/rebuild_mondb.cc (hunk 10 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: if we have an OSD with osdmap[4..10] first, and

1. case 1, we update our store with osdmap[20..28], then osdmap[4..20) is chopped. no old maps is left in store. and we now have osdmap[20..28]
1. case 2, we update our store with osdmap[6..20], then osdmap[4..6) is chopped. no old maps is left in store. and we now have osdmap[6..20].
1. case 3, we update our store with osdmap[1..3], try to trim osdmap[4..1], so no maps gets trimmed. no maps is added.
1. case 4, we update our store with osdmap[1..6], try to trim osdmap[4..1], so no maps gets trimmed, and no maps is added.

so i think we are good?

File: src/tools/rebuild_mondb.cc (hunk 8 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: see https://github.com/ceph/ceph/blob/master/src/osd/OSD.cc#L6824, superblock and new osdmaps are written in the same transaction. so OSD can hardly have "new maps but an out-of-date newest_map". am i right?

File: src/tools/rebuild_mondb.cc (hunk 7 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: @gregsfortytwo can we leave the refactor for another PR? as i need to restructure monitor carefully to cater the needs on both sides.

File: doc/rados/troubleshooting/troubleshooting-mon.rst (hunk 6 above)
Comment (tchaikov) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: @athanatos but I'm not sure if you handled that.

i do handle this. as @gregsfortytwo pointed out. see https://github.com/ceph/ceph/pull/10933#discussion-diff-77090613R217

Comment (tchaikov) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: @gregsfortytwo 
> I'd assumed it would be extracting to some intermediate format and then turning them into the mon format at the end.

i also evaluated this approach, it can be executed in parallel when collecting the OSD side info, hence more efficient. but seems current way is simpler. and this fix is only for disaster recovery which is not supposed to be used very frequently. so i am inclined to use this iterative way.

**Comment by tchaikov** [comment]
all comments addressed except for https://github.com/ceph/ceph/pull/10933#discussion_r77090965 and https://github.com/ceph/ceph/pull/10933#discussion_r77090792.

i am working on a ceph-qa-suite task to 

1. nuke all store.db
1. rebuild one from OSDs
1. restore it to the first mon, mkfs on other mons
1. revive them

**Comment by tchaikov** [comment]
jenkins, retest this please.

**Comment by tchaikov** [comment]
changelog

- rebased against master
- ceph-monstore-tool: instead of getting the client.admin key and add an almighty keyring, import all keyrings from the `--keyring` file, to ready the mon.

**Comment by tchaikov** [comment]
test added at https://github.com/ceph/ceph-qa-suite/pull/1169

File: src/tools/rebuild_mondb.cc (hunk 8 above)
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Yeah, if we assume the OSD isn't messed up, than this is true.

File: src/tools/rebuild_mondb.cc [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+      cerr << "missing " << oid << std::endl;
+      return -EINVAL;
+    }
+    t->put(prefix, ms.combine_strings("full", e), bl);
+  }
+  if (sb.newest_map > last_committed) {
+    t->put(prefix, last_committed_name, sb.newest_map);
+    t->put(prefix, ms.combine_strings("full", "latest"), sb.newest_map);
+  }
+  ms.apply_transaction(t);
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Hmm, can we actually write out all of the OSDMaps in one transaction?

Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: no, will write them in batch.

**Comment by tchaikov** [comment]
retest this please.

**Comment by tchaikov** [comment]
changelog

- write osdmaps in batch of size 30
- use `#pragma once` for guarding multiple inclusion of `rebuild_mondb.h`

**Comment by tchaikov** [comment]
@athanatos test passed at http://pulpito.ceph.com/kchai-2016-09-09_07:30:03-rados:singleton-wip-kefu-testing---basic-mira/

**Comment by tchaikov** [comment]
retest this please.

**Comment by zphj1987** [comment]
can this be test now 
@tchaikov

**Comment by tchaikov** [comment]
@zphj1987 if you'd like to give it a spin, i am happy to learn if it works for you. =)

**Comment by zphj1987** [comment]
@tchaikov 
i will test it in my environment, and tell you the result ，good job!

**Comment by tchaikov** [comment]
changelog

- rebase against master to resolve conflicts.

@athanatos could you take a look again? thanks!

File: src/tools/rebuild_mondb.cc [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+  const string first_committed_name("first_committed");
+  const string last_committed_name("last_committed");
+  epoch_t first_committed = ms.get(prefix, first_committed_name);
+  epoch_t last_committed = ms.get(prefix, last_committed_name);
+
+  // trim stale maps
+  unsigned ntrimmed = 0;
+  {
+    auto t = make_shared<MonitorDBStore::Transaction>();
+    for (auto e = first_committed; e && e < sb.oldest_map; e++) {
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Why e && e < oldest_map? What does first_committed = 0 mean?

File: src/tools/rebuild_mondb.cc [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+      t->erase(prefix, e);
+      t->erase(prefix, ms.combine_strings("full", e));
+      ntrimmed++;
+    }
+    if (sb.oldest_map > first_committed) {
+      t->put(prefix, first_committed_name, sb.oldest_map);
+    }
+    ms.apply_transaction(t);
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Don't really need to apply if it empty, right -- move into immediately prior if block?

File: src/tools/rebuild_mondb.cc [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+  }
+
+  // this number comes from the default value of osd_target_transaction_size,
+  // so we won't OOM or stuff too many maps in a single transaction if OSD is
+  // keeping a large series of osdmap
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: If I'm reading this right, we could end up with an incremental map without the corresponding full map, and we can end up with maps outside of [first_committed_name, last_committed_name].  You probably want to make sure that each transaction leaves the store consistent.

File: src/tools/rebuild_mondb.cc (hunk 13 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: will do.

File: src/tools/rebuild_mondb.cc (hunk 12 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: if there is not such a k/v pair in store.db, `ms.get(prefix, key)` will simply return 0. this creates an illusion for ceph-objectstore-tool that there is always an osdmap#0 in store.db even there is nothing in it. and in real world, the osdmap starts at 1. i will accompany the code with some comments.

File: src/tools/rebuild_mondb.cc (hunk 14 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: will do.

**Comment by tchaikov** [comment]
changelog

- add comment explaining why we don't want to trim osdmap#0.
- do not apply transaction if it's empty.
- make sure the store.db is consistent after each transaction adding osdmaps.

@athanatos mind taking a look again? thanks as always.

File: src/tools/rebuild_mondb.cc [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+
+  // trim stale maps
+  unsigned ntrimmed = 0;
+  // osdmap starts at 1. if we have a "0" first_committed, then there is nothing
+  // to trim. and "1 osdmaps trimmed" in the output message is misleading. so
+  // let's make it an exception.
+  for (auto e = first_committed; first_committed && e < sb.oldest_map; e++) {
+    t->erase(prefix, e);
+    t->erase(prefix, ms.combine_strings("full", e));
+    t->put(prefix, first_committed_name, e + 1);
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Probably do this once at the end?

File: src/tools/rebuild_mondb.cc [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+  epoch_t first_committed = ms.get(prefix, first_committed_name);
+  epoch_t last_committed = ms.get(prefix, last_committed_name);
+  auto t = make_shared<MonitorDBStore::Transaction>();
+
+  // trim stale maps
+  unsigned ntrimmed = 0;
+  // osdmap starts at 1. if we have a "0" first_committed, then there is nothing
+  // to trim. and "1 osdmaps trimmed" in the output message is misleading. so
+  // let's make it an exception.
+  for (auto e = first_committed; first_committed && e < sb.oldest_map; e++) {
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Seems like you can avoid the first_commited check if you change the loop to go up to MIN(sb.oldest_map, last_committed)?

File: src/tools/rebuild_mondb.cc [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+      if (nread <= 0) {
+        cerr << "missing " << oid << std::endl;
+        return -EINVAL;
+      }
+      t->put(prefix, ms.combine_strings("full", e), bl);
+    }
+    nadded++;
+
+    // last_committed
+    t->put(prefix, last_committed_name, e);
```
Comment (athanatos) [inline src/tools/rebuild_mondb.cc]: Put this into the if block with the apply?

Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: then i have to repeat them in the 2nd if block with the apply.

File: src/tools/rebuild_mondb.cc (hunk 15 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: will do.

File: src/tools/rebuild_mondb.cc (hunk 16 above)
Comment (tchaikov) [inline src/tools/rebuild_mondb.cc]: - if `MIN(sb.oldest_map, last_committed)` evaluates to `sb.oldest_map`, we should trim osdmap[first_committed, sb.oldest_map**)**, 
- if `MIN(sb.oldest_map, last_committed)` evaluates to `last_committed`, we should trim osdmap[first_committed, last_committed**]**, 

blindly trim up to sb.oldest_map might be easier to understand.

**Comment by tchaikov** [comment]
changelog

- update the "first_committed" once in the end.

**Comment by tchaikov** [comment]
changelog

- remove trailing spaces.

@athanatos

**Comment by tchaikov** [comment]
being tested at http://pulpito.front.sepia.ceph.com:80/kchai-2016-09-12_18:18:40-rados:singleton-wip-kefu-testing---basic-mira/

**Comment by athanatos** [comment]
LGTM, @gregsfortytwo One last look?

File: doc/rados/troubleshooting/troubleshooting-mon.rst (hunk 6 above)
Comment (jecluis) [inline doc/rados/troubleshooting/troubleshooting-mon.rst]: As mentioned to @tchaikov on IRC, it bothers me that the admin has to copy the osd data from each osd to the node where recovery is happening. This seems slow and unnecessary. We could simply grab the maps from `current/meta/` from all osds.

And, even though this is not something we expect to happen often (ideally, at all), I can imagine how an admin would be closing to tears if he had to copy every. single. osd. to get to the maps during DR. Poor admin. :(

**Comment by tchaikov** [comment]
per the discussion with @jecluis 

* import from OSD's keyring also
* check the crc of osdmaps
* consider rsync'ing and using the meta directory only.

**Comment by tchaikov** [comment]
changelog

- check the crc of osdmaps before putting them into store.db.

**Comment by tchaikov** [comment]
test passed at http://pulpito.ceph.com/kchai-2016-09-15_15:48:11-rados:singleton-wip-kefu-testing---basic-mira/ .

@jecluis @athanatos @gregsfortytwo, thanks for your reviews.

is this changeset good to merge?

**Comment by athanatos** [comment]
It looks good to me, it would probably be good to get a final ack from @gregsfortytwo .
